dbms/src/Flash/Executor/PipelineExecutorStatus.cpp

```diff
@@ -92,15 +92,39 @@ void PipelineExecutorStatus::onEventSchedule() noexcept
 
 void PipelineExecutorStatus::onEventFinish() noexcept
 {
-    std::lock_guard lock(mu);
-    assert(active_event_count > 0);
-    --active_event_count;
-    if (0 == active_event_count)
-        cv.notify_all();
+    bool query_finished = false;
+    {
+        std::lock_guard lock(mu);
+        assert(active_event_count > 0);
+        --active_event_count;
+        if (0 == active_event_count)
+        {
+            cv.notify_all();
+            query_finished = true;
+        }
+    }
+    if (query_finished)
+    {
+        if (auto ret = result_queue_holder.tryGet(); ret)
+            (*ret)->finish();
+    }
 }
 
 void PipelineExecutorStatus::cancel() noexcept
 {
     is_cancelled.store(true, std::memory_order_release);
+
+    if (auto ret = result_queue_holder.tryGet(); ret)
+        (*ret)->cancel();
+}
+
+ResultQueuePtr PipelineExecutorStatus::registerResultQueue(size_t queue_size) noexcept
+{
+    auto result_queue = std::make_shared<ResultQueue>(queue_size);
+    ResultQueuePtr tmp = result_queue;
+    result_queue_holder.set(std::move(tmp));
+    if unlikely (is_cancelled)
```

Comment:
Is it necessary to check `is_cancelled` here?

Reply:
Consider the situation
```
threadA: exec_status.cancel --> if result_queue exist call result_queue.cancel
threadB: create result_queue
```
if the `result_queue` created after `exec_status.cancel` and before `result_queue.cancel`, and then `result_queue.cancel` is never called.

Reply:
> Is it necessary to check `is_cancelled` here?

As for https://github.com/pingcap/tiflash/pull/7175/files#diff-df63e0df429584d43e733e183d00ee6fba027e9578257df168268444911c2211R93-R95
`result_queue.cancel` shouldn't be called, removed.